node/pkg/watchers/evm/finalizers/optimism.go

```diff
@@ -1,95 +1,151 @@
 
-type FinalizerEntry struct {
-	l2Block uint64
-	l1Block uint64
-}
 type OptimismFinalizer struct {
 	logger                 *zap.Logger
 	connector              connectors.Connector
 	l1Finalizer            interfaces.L1Finalizer
-	latestFinalizedL2Block uint64
+	latestFinalizedL2Block big.Int
 
-	// finalizerMapping is a array of FinalizerEntry structs with the L2 block number that has been verified and its corresponding L1 block number
-	finalizerMapping []FinalizerEntry
+	// finalizerMapping is a array of RollupInfo structs with the L2 block number that has been verified and its corresponding L1 block number
+	finalizerMapping []RollupInfo
+
+	// These are used for querying the ctc contract.
+	ctcRawClient *ethRpc.Client
+	ctcClient    *ethClient.Client
+
+	// This is used to grab the rollup information from the ctc contract
+	ctcCaller *ctcAbi.OptimismCtcAbiCaller
 }
 
-func NewOptimismFinalizer(ctx context.Context, logger *zap.Logger, connector connectors.Connector, l1Finalizer interfaces.L1Finalizer) *OptimismFinalizer {
-	return &OptimismFinalizer{
+func NewOptimismFinalizer(
+	ctx context.Context,
+	logger *zap.Logger,
+	connector connectors.Connector,
+	l1Finalizer interfaces.L1Finalizer,
+	ctcChainUrl string,
+	ctcChainAddress string,
+) (*OptimismFinalizer, error) {
+
+	ctcRawClient, err := ethRpc.DialContext(ctx, ctcChainUrl)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create raw client for url %s: %w", ctcChainUrl, err)
+	}
+
+	ctcClient := ethClient.NewClient(ctcRawClient)
+
+	addr := ethCommon.HexToAddress(ctcChainAddress)
+
+	ctcCaller, err := ctcAbi.NewOptimismCtcAbiCaller(addr, ctcClient)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create ctc caller for url %s: %w", ctcChainUrl, err)
+	}
+
+	finalizer := &OptimismFinalizer{
 		logger:                 logger,
 		connector:              connector,
 		l1Finalizer:            l1Finalizer,
-		latestFinalizedL2Block: 0,
-		finalizerMapping:       make([]FinalizerEntry, 0),
+		latestFinalizedL2Block: *big.NewInt(0),
+		finalizerMapping:       make([]RollupInfo, 0),
+		ctcRawClient:           ctcRawClient,
+		ctcClient:              ctcClient,
+		ctcCaller:              ctcCaller,
 	}
+
+	return finalizer, nil
+}
+
+// Both types are big.Int because that is what the abi returns.
+// However, there are 2 points to note:
+// The L1 block number is a uint40 in the abi.  So, safe to convert to Uint64.
+// The L2 block number is a uint256 in the abi.  So, this is never converted to anything else.
+type RollupInfo struct {
+	l2Block big.Int
+	l1Block big.Int
+}
+
+func (f *OptimismFinalizer) GetRollupInfo(ctx context.Context) (RollupInfo, error) {
+	// Get the current latest blocks.
+	opts := &ethBind.CallOpts{Context: ctx}
+	var entry RollupInfo
+	l1Block, err := f.ctcCaller.GetLastBlockNumber(opts)
+	if err != nil {
+		return entry, fmt.Errorf("failed to get L1 block: %w", err)
+	}
+	l2Block, err := f.ctcCaller.GetTotalElements(opts)
```

Comment:
Is there a risk of a race condition between these 2 function calls and a contract update?  For example we call `GetLastBlockNumber` and get the L1 block, then the CTC is updated on-chain, then we call `GetTotalElements` for the L2 block.  Now a newer L2 block is associated with an older L1 block potentially leading to guardiand thinking an L2 block is finalized when it hasn't actually been finalized.

Maybe we should do these in reverse so that in the worst case an older L2 block is associated with a newer L1 block.

Reply:
Changed